File: typescript/lld/marketplace/cab-booking/DemandSupplySurge.ts
import { SurgeStrategy } from './SurgeStrategy';

export class DemandSupplySurge implements SurgeStrategy {
    multiplier(openRequests: number, availableDrivers: number): number {
        if (availableDrivers === 0) return 3.0;              // cap even with zero supply
        const ratio = openRequests / availableDrivers;
        return Math.min(3.0, Math.max(1.0, ratio));           // 1.0x floor, 3.0x cap
    }
}

File: python/lld/marketplace/cab-booking/demand_supply_surge.py
from surge_strategy import SurgeStrategy


class DemandSupplySurge(SurgeStrategy):
    def multiplier(self, open_requests: int, available_drivers: int) -> float:
        if available_drivers == 0:
            return 3.0                          # cap even with zero supply
        ratio = open_requests / available_drivers
        return min(3.0, max(1.0, ratio))          # 1.0x floor, 3.0x cap